### ISSUE TITLE
System Cleaner Still Scans Excluded Paths

### ISSUE THREAD

**Issue by Yowlen** [d4rken-org/sdmaid#975]
This is more of a performance issue than anything else. Ultimately, items from excluded paths don't get added to the list of stuff to clean, but if the excluded folders have a lot of files in them, the initial scan can take longer than it needs to be.

Easiest to reproduce by first timing the scan length, then creating a ton of files in a folder on the SD card, excluding that folder from System Cleaner, and timing the scan again.

**Comment by d4rken** [comment]
The SystemCleaner starts it's search at the root of storages, emits all found items and runs the filters against them. The exclusions are then applied to each found item, which doesn't keep them from being traversed as you noticed.

What you want is that exclusions stop the SystemCleaner from traversing specific directories. This would require that we pass the exclusions down to lower levels of SD Maids tooling. We would have to make sure that the shell based search (using find), the JavaFile and the DocumentFile based search honors these exclusions. This would be possible for the last two, but for the shell based search it's a bit more difficult.

There is the issue that the command line buffer may be limited on some devices, so we might not be able to append many exclusions to the search command without causing unforseen issues, and SD Maid supports simple as well as regex based exclusions. Making sure find excludes directories from being traversed, based on regular expressions, also seems difficult.

**Comment by d4rken** [comment]
Not sure this would be worth it.

**Comment by Yowlen** [comment]
I figured it would require something of that nature. If you don't care to do it, I understand. I figure it would be worth asking about, regardless.

And yeah, I figured incorrect exclusions, both regex or otherwise, would end up messing up the filters. But as long as it only affects the scans, I don't see the issue. It's not like it would actually delete anything, only exclude stuff from being deleted.

And if the system's already working to find what gets deleted, I don't see an issue with moving it further low-level. Just gotta make sure the syntax is correct, I'd imagine.

Just something to think about, at least, and I'd be willing to test if you do decide to do it.
